### PR TITLE
Add Support for Cluster Proxy

### PR DESCRIPTION
**Pull request by VaishnaviHire**
## How Has This Been Tested?
- OCP cluster 4.11
- Update `cluster` Proxy resource to resemble -
```
apiVersion: config.openshift.io/v1
kind: Proxy
metadata:
  name: cluster
spec:
  httpProxy: http://<username>:<pswd>@<ip>:<port> 
  httpsProxy: http://<username>:<pswd>@<ip>:<port> 
  noProxy: example.com 
  readinessEndpoints:
  - http://www.google.com 
  - https://www.google.com
  trustedCA:
    name: user-ca-bundle 
```
- Deploy live-build : quay.io/modh/rhods-operator-live-catalog:1.20.0-rhods-5811
- Start Jupyter Server
- Verify:
      -  if `trusted-ca` configmap is created in user namespace
      - configmap is mounted in Notebook resource
      - proxy Env variables are added to Notebook pod
## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 1]
```diff
@@ -253,8 +265,103 @@ func (w *NotebookWebhook) Handle(ctx context.Context, req admission.Request) adm
...
+					return fmt.Errorf("notebook image container not found %v", notebook.Name)
+				}
+
+			}
+			break
+		}
+	}
+
+	// Add trusted-ca volume
+	notebookVolumes := &notebook.Spec.Template.Spec.Volumes
```
Comment (Jooho) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:337]: The first question I have is "What is this tls-certificate volume?" 
 "https://github.com/opendatahub-io/kubeflow/blob/24495240d0e896124468509d082e2d57d767c2c8/components/odh-notebook-controller/controllers/notebook_webhook.go#L198-L224

There is no volume mount for this volume in the notebook container.

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 2]
```diff
@@ -253,8 +265,103 @@ func (w *NotebookWebhook) Handle(ctx context.Context, req admission.Request) adm
...
+	err := w.Client.List(context.TODO(), proxyResourceList)
+	if err != nil {
+		return false
+	}
+
+	for _, proxy := range proxyResourceList.Items {
+		if proxy.Name == "cluster" {
+			if proxy.Spec.HTTPProxy != "" && proxy.Spec.HTTPSProxy != "" &&
+				proxy.Spec.NoProxy != "" && proxy.Spec.TrustedCA.Name != "" {
+				return true
```
Comment (Jooho) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:279]: For non-ssl proxy servers, TrustedCA should not be set. so it should check the TrustedCA value independently.

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 1 above)
Comment (Jooho) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:337]: VolumeMount must be added to notebook container and it looks like the following:
~~~
        - name: custom-ca
          volumeMounts:
          - name: trusted-ca
            mountPath: /etc/pki/ca-trust/extracted/pem
            readOnly: true
~~~

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 3]
```diff
@@ -253,8 +265,103 @@ func (w *NotebookWebhook) Handle(ctx context.Context, req admission.Request) adm
...
+				for _, env := range imgContainer.Env {
+					if key == env.Name {
+						keyExists = true
+					}
+				}
+				if !keyExists {
+					newVars = append(newVars, corev1.EnvVar{Name: key, Value: val})
+				}
+			}
+
```
Comment (Jooho) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:314]: Does the value come from Proxy cluster? it does set ""
~~~
        - name: HTTPS_PROXY
        - name: HTTP_PROXY
        - name: NO_PROXY
~~~

**Comment by Jooho** [comment]
Test failed.

Issues:
1. proxy environments were not updated.
2. PIP_CERT env variable should be optional. This has to be set when TrustCA is set.
3. No volumnMount exists for the custom certificate in notebook container.
4. The current check logic for the cluster wide proxy always checks TrustCA but for non-ssl proxy server does not need it. So it should separate it from the current logic.

**Review by Jooho — APPROVED** [review]
/lgtm

**Comment by Jooho** [comment]
@LaVLaS could you review this ?

**Review by LaVLaS — APPROVED** [review]
/lgtm

**Comment by shalberd** [comment]
Very nice, cool to see this, as it is a big part of disconnected / airgapped / enterprise support :-)

**Comment by Jooho** [comment]
@shalberd hope to see many people use this feature :)

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 4]
```diff
@@ -253,8 +265,135 @@ func (w *NotebookWebhook) Handle(ctx context.Context, req admission.Request) adm
...
+	for _, proxy := range proxyResourceList.Items {
+		if proxy.Name == "cluster" {
+			if proxy.Status.HTTPProxy != "" && proxy.Status.HTTPSProxy != "" &&
+				proxy.Status.NoProxy != "" {
+				// Update Proxy Env variables map
+				proxyEnvVars["HTTP_PROXY"] = proxy.Status.HTTPProxy
+				proxyEnvVars["HTTPS_PROXY"] = proxy.Status.HTTPSProxy
+				proxyEnvVars["NO_PROXY"] = proxy.Status.NoProxy
+				if proxy.Spec.TrustedCA.Name != "" {
+					proxyEnvVars["PIP_CERT"] = "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem"
```
Comment (shalberd) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:284]: this is too specific to PIP_CERT or pip. The trusted CAs also need to be available for htttp requests from python to custom CA-based Http locations, e.g. Artifactory or Gitlab Servers running with server certs based on a custom CA. In Python, that is I think env var REQUESTS_CA_BUNDLE.

**Pull request merged**